examples/pointer_theme_window.rs

```diff
@@ -0,0 +1,491 @@
+            self.keyboard_focus = true;
+        }
+    }
+
+    fn leave(
+        &mut self,
+        _: &Connection,
+        _: &QueueHandle<Self>,
+        _: &wl_keyboard::WlKeyboard,
+        surface: &wl_surface::WlSurface,
+        _: u32,
+    ) {
+        if self.window.as_ref().map(Window::wl_surface) == Some(surface) {
+            println!("Release keyboard focus on window");
+            self.keyboard_focus = false;
+        }
+    }
+
+    fn press_key(
+        &mut self,
+        _conn: &Connection,
+        _qh: &QueueHandle<Self>,
+        _: &wl_keyboard::WlKeyboard,
+        _: u32,
+        event: KeyEvent,
+    ) {
+        println!("Key press: {:?}", event);
+    }
+
+    fn release_key(
+        &mut self,
+        _: &Connection,
+        _: &QueueHandle<Self>,
+        _: &wl_keyboard::WlKeyboard,
+        _: u32,
+        event: KeyEvent,
+    ) {
+        println!("Key release: {:?}", event);
+    }
+
+    fn update_modifiers(
+        &mut self,
+        _: &Connection,
+        _: &QueueHandle<Self>,
+        _: &wl_keyboard::WlKeyboard,
+        _serial: u32,
+        modifiers: Modifiers,
+    ) {
+        println!("Update modifiers: {:?}", modifiers);
+    }
+}
+
+impl PointerHandler for SimpleWindow {
+    fn pointer_frame(
+        &mut self,
+        _conn: &Connection,
+        _qh: &QueueHandle<Self>,
+        _pointer: &wl_pointer::WlPointer,
+        events: &[PointerEvent],
+    ) {
+        use PointerEventKind::*;
+        for event in events {
+            // Ignore events for other surfaces
+            if Some(&event.surface) != self.window.as_ref().map(Window::wl_surface) {
+                continue;
+            }
+            match event.kind {
+                Enter { .. } => {
+                    println!("Pointer entered @{:?}", event.position);
+                }
+                Leave { .. } => {
+                    println!("Pointer left");
+                }
+                Motion { .. } => {}
+                Press { button, .. } => {
+                    println!("Press {:x} @ {:?}", button, event.position);
+                    self.shift = self.shift.xor(Some(0));
+                }
+                Release { button, .. } => {
+                    println!("Release {:x} @ {:?}", button, event.position);
+                }
+                Axis { horizontal, vertical, .. } => {
+                    println!("Scroll H:{:?}, V:{:?}", horizontal, vertical);
+                }
+            }
+        }
+    }
+}
+
+impl ShmHandler for SimpleWindow {
+    fn shm_state(&mut self) -> &mut ShmState {
+        &mut self.shm_state
+    }
+}
+
+impl SimpleWindow {
+    pub fn draw(&mut self, conn: &Connection, qh: &QueueHandle<Self>) {
+        if let Some(window) = self.window.as_ref() {
+            if self.themed_pointer.is_none() && self.pointer.is_some() {
+                let ptr = self.pointer.clone().unwrap();
+                let ptr_data: &PointerData = ptr.data().unwrap();
```

Comment:
Couldn't we return both the ThemedPointer instance and the WlPointer at the same time? Not very fun to have to poke into the user data all the time.

This seems to force a pointer when created to have a theme with it, which for most cases is fine. But sometimes it may be desired to not touch the cursor theme at all.

Reply:
> Couldn't we return both the ThemedPointer instance and the WlPointer at the same time? Not very fun to have to poke into the user data all the time.

Ok, I see what you mean. I think I can improve that.

> This seems to force a pointer when created to have a theme with it, which for most cases is fine. But sometimes it may be desired to not touch the cursor theme at all.

If I understand your question correctly, you mean like in the case of an application which wants to set the cursor image themselves and doesn't want to use the ThemedPointer to set the cursor at all? `get_pointer` is available which just sets the ThemedPointer to be default, and I don't believe it will have an effect if it isn't used, but I think I could probably put it behind a feature so it's left out completely if the user doesn't want to use it.

Reply:
Yes